.pre-commit-config.yaml

```diff
@@ -10,6 +10,7 @@ repos:
           - id: end-of-file-fixer
             exclude: '\.ipynb$'
           - id: check-toml
+            exclude: '{{cookiecutter.repo_name}}/pyproject.toml'
```

Comment:
I guess you're excluding this, as we now use some templating in the file, correct?

To still make sure that the generated files pass our own quality standards, we should extend the tests with one variation that uses ruff.

Reply:
From my understanding, the [check-toml pre-commit hook](https://github.com/pre-commit/pre-commit-hooks/blob/main/pre_commit_hooks/check_toml.py) tries to parse the toml within the project directory as-is, which leads to a flag because it still contains template strings. This is not necessarily related to ruff/pylint, but to the way the check itself is executed. Let's discuss how to solve this :)

Reply:
Following up our discussion, I have implemented that the tests now execute the pre-commit hooks. On top of that, I've added the option to specify different configurations to run the tests in. With these changes in place, I'd suggest to:
1. Exclude everything on "{{cookiecutter.repo_name}}" from the top-level pre-commit configuration, as these are now validated in the tests (after cookiecutter initialized the repo)
2. Add some configuration with Ruff to our [permutations](https://github.com/HotSprings-GmbH/data-science-project-template/blob/817b581fd6aa5109826a975e9e87f777185b642f/tests/test_cookiecutter_template.py#L82).

Reply:
1. Done in a979095
2. Done in fe346c3. Two comments: 

* I took the liberty of adding separate tests for validating the contents of the python environment + jupyter & linter separately. I also refactored the validate gitlab function to be more generic and easily extendable if we want to integrate GitHub / Azure DevOps sometime in the future 
* Instead of hardcoding a set of configurations, I implemented a function that selects N random permutations of all possible configurations based on cookiecutter.json. See function [get_all_possible_configuration_permutations()](https://github.com/HotSprings-GmbH/data-science-project-template/pull/28/files#diff-7fdf92eceaf9d2945d142a8d819b281670b61c82f3ef995557228343075b6ab8R35) in tests/test_cookiecutter_template.py

Reply:
Looks good!

Not sure if I like the approach of choosing permutations at random. I think we want to have reproducible build and make sure that we cover each option at least once, right? However, we also don't want to test all permutations in the interest of keeping build times low... Let discuss to find a good tradeoff :)